# Process

To save operation reseouces, the project only assigned sentiment scores to the corresponding sentence indices when saving LLM's results. Accordingly, the states list and sentences were merged into newly retrieved dataset with sentiment scores.

In [ ]:
import pandas as pd
import ast
import itertools
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import precision_recall_fscore_support,precision_score, recall_score, f1_score,accuracy_score
import networkx as nx
import netbone as nb
from netbone.filters import threshold_filter

In [ ]:
scores = pd.read_csv("../data/layer1/layer1_sentiment.csv")
sentences = pd.read_csv("../data/layer1/layer1_sentence.csv")

sentences["index"] = sentences.index
merged = pd.merge(scores, sentences, on='index', how="left")

In [8]:
# inspect the data
print(merged.info())
print(merged.head(3))
merged.to_csv("../data/layer1/layer1_merged.csv",encoding="utf-8",index=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38688 entries, 0 to 38687
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   index      38688 non-null  int64 
 1   sentiment  38688 non-null  int64 
 2   sentence   38688 non-null  object
 3   states     38688 non-null  object
dtypes: int64(2), object(2)
memory usage: 1.2+ MB
None
   index  sentiment                                           sentence  \
0      0         -1  Ukraine war briefing: Zelenskyy vows his count...   
1      1          1  Ukraine’s President Volodymyr Zelenskyy has vo...   
2      2         -1  Ukraine has been allowing Russia to pipe gas t...   

                  states  
0  ['Russia', 'Ukraine']  
1  ['Russia', 'Ukraine']  
2  ['Russia', 'Ukraine']  


# Sample
According to the Cochran's formula, the project sampled 269 items ((e=0.05, p=0.5, Z=1.645/90% CI, N=38688)) from the dataset, and categorized the samples into Positive, Negative and Neutral sentiments manually. By comparing the results of the human annotation with those of the LLM, a 3*3 multi-class confusion matrix was generated.

Furthermore, the project also calculated Precision, Recall and F1 for each category and the overall performance. Take the positive category as an example, 

TP: Sentences classified as positive by both the LLM and the human annotator.
FP: Sentences classified as positive by the LLM but identified as non-positive by the human.
TN: Sentences classified as non-positive by both the LLM and the human.
FN: Sentences classified as non-positive by the LLM but identified as positive by the human.

$$Precision = \frac{TP}{TP + FP}$$

$$Recall = \frac{TP}{TP + FN}$$

$$F1 = 2 * \frac{Precision * Recall}{Precision + Recall}$$

The metrices for the overall model was calculated through the arithmetic mean of the three categories.

In [7]:
sample = merged.sample(n=269, random_state=22)
sample.to_csv("../data/sample/sentiment_sample.csv",encoding="utf-8",index=False)

In [9]:
# Assessment
annotation = pd.read_csv("../data/sample/sentiment_sample_annotated.csv")

# under each category
y_true = annotation["human"]
y_pred = annotation["sentiment"]

labels = [-1, 0, 1]
precision, recall, f1, _ = precision_recall_fscore_support(y_true, y_pred, average=None, labels=[-1, 0, 1])
accs = []
for i, label in enumerate(labels):
    idx = (y_true == label)
    acc = accuracy_score(y_true[idx], y_pred[idx])
    accs.append(acc)
    print(f"Sentiment {label}: Precision={precision[i]:.3f}, Recall={recall[i]:.3f}, F1={f1[i]:.3f}, Accuracy={acc:.3f}")

Sentiment -1: Precision=0.915, Recall=0.831, F1=0.871, Accuracy=0.831
Sentiment 0: Precision=0.792, Recall=0.800, F1=0.796, Accuracy=0.800
Sentiment 1: Precision=0.600, Recall=0.794, F1=0.684, Accuracy=0.794


In [11]:
# overall
overall_precision = precision_score(y_true, y_pred, average="macro")
overall_recall = recall_score(y_true, y_pred, average="macro")
overall_f1 = f1_score(y_true, y_pred, average="macro")
overall_acc = accuracy_score(y_true, y_pred)

print(f"Overall Precision: {overall_precision:.3f}")
print(f"Overall Recall: {overall_recall:.3f}")
print(f"Overall F1: {overall_f1:.3f}")
print(f"Overall Accuracy: {overall_acc:.3f}")

Overall Precision: 0.769
Overall Recall: 0.808
Overall F1: 0.784
Overall Accuracy: 0.814


# Sentiment Aggregation
In this stage, sentiment scores were aggregated at the level of specific state-pairs based on the sentences where they appeared.

In [ ]:
merged = pd.read_csv("../data/layer1/layer1_merged.csv")

def parse_states(s):
    try:
        return ast.literal_eval(s)
    except Exception:
        return []

merged['states_list'] = merged['states'].apply(parse_states)

# generate pairs
pairs = []
for _, row in merged.iterrows():
    countries = row['states_list']
    sentiment = row['sentiment']
#     if not isinstance(countries, list) or len(countries) < 2:
#         continue
    # sort to avoid repeating the same states
    uniq = sorted(set(countries))
    for a, b in itertools.combinations(uniq, 2):
        pairs.append((a, b, sentiment))

pair_df = pd.DataFrame(pairs, columns=['s1', 's2', 'score'])
pair_stats = pair_df.groupby(['s1', 's2'], as_index=False).agg(
    total_score=('score', 'sum'),
    mention_count=('score', 'count')
)
pair_stats['avg_score'] = pair_stats['total_score'] / pair_stats['mention_count']
print(pair_stats.head(3))
pair_stats.to_csv('../data/layer1/layer1_pair_scores_all.csv', encoding='utf-8', index=False)

            s1                        s2  total_score  mention_count  \
0  Afghanistan                   Bahrain            0              1   
1  Afghanistan                   Belarus           -3              4   
2  Afghanistan  Central African Republic           -1              1   

   avg_score  
0       0.00  
1      -0.75  
2      -1.00  


##  Extraction
The study extracted the network's backbone using the Disparity Filter Algorithm (implemented via the netbone library). This algorithm operates on the null hypothesis that a node’s total weight is distributed randomly among its $k$ neighbors. Under this assumption, the probability of a specific tie carrying a certain weight fraction $x$ follows the distribution of order statistics. If the calculated p-value ($P$) is below the significance level $\alpha$, the tie is considered statistically significant rather than a result of random fluctuations, representing a functional or intentional connection.

Assume $x$ represents the fraction of a node's total weight allocated to a specific neighbor. Based on the Order Statistics of a random split, the probability density function is:$$f(x) = (k-1)(1-x)^{k-2}$$The p-value $P$, representing the probability of observing a weight fraction $\ge x$ by chance, is:$$P = \int_{x}^{1} (k-1)(1-u)^{k-2} du = (1-x)^{k-1}$$

To achieve the goal of preserving maximum nodes while filtering out stochastic links, the study set different thresholds for positive and negative ties. For positive ties, the threshold was set at 0.2 to retain substantial links below baseline random expectation. Conversely, for negative ties that serve as an exclusionary filter for another layer, a more stringent threshold of 0.05 was employed. This rigorous standard minimizes the risk of misidentifying random negative interactions as adversarial or hostile.

In [24]:
pair_scores = pd.read_csv('../data/layer1/layer1_pair_scores_all.csv')

In [59]:
df_input = pair_scores.rename(columns={'s1': 'source', 's2': 'target','mention_count':'weight'})
df_with_alpha = nb.disparity(df_input, weight='weight')

backbone_positive = threshold_filter(df_with_alpha, 0.2)
backbone_negative = threshold_filter(df_with_alpha, 0.05)

assert backbone_positive
assert backbone_negative
backbone_positive_df = nx.to_pandas_edgelist(backbone_positive)
backbone_negative_df = nx.to_pandas_edgelist(backbone_negative)

backbone_positive_df = backbone_positive_df.rename(columns={'source': 's1', 'target': 's2'})
backbone_negative_df = backbone_negative_df.rename(columns={'source': 's1', 'target': 's2'})

backbone_positive_merged = pd.merge(
    backbone_positive_df, 
    pair_scores[['s1', 's2', 'avg_score']], 
    on=['s1', 's2'], 
    how='left'
)

backbone_negative_merged = pd.merge(
    backbone_negative_df, 
    pair_scores[['s1', 's2', 'avg_score']], 
    on=['s1', 's2'], 
    how='left'
)

backbone_positive_filtered = backbone_positive_merged[backbone_positive_merged['avg_score'] > 0]
backbone_negative_filtered = backbone_negative_merged[backbone_negative_merged['avg_score'] < 0]

backbone_positive_df.to_csv('../data/layer1/layer1_backbone_pos.csv', encoding='utf-8', index=False)
backbone_negative_df.to_csv('../data/layer1/layer1_backbone_neg.csv', encoding='utf-8', index=False)

Disparity Filter
Disparity Filter
